Memory leak detected on Linux builds

The memory sanitizer tool did not report leaks in MacOS, but it is reporting in Linux builds.

Make all stops with:
``` 
rm -rf build
mkdir -p build && \
        gcc -DTEST -fsanitize=address -g -O0 -o build/str_join str_join.c && \
        gcc -DTEST -fsanitize=address -g -O0 -o build/str_split str_split.c && \
        gcc -DTEST -fsanitize=address -g -O0 -o build/functools functools.c && \
        ./build/str_join && ./build/str_split && ./build/functools
test_join - Passed

=================================================================
==5735==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 9 byte(s) in 1 object(s) allocated from:
    #0 0xffffa1a7a2f4 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0xaaaac5a81774 in str_join /home/marciorj/Documents/projects/libfunctools/str_join.c:26
    #2 0xaaaac5a820cc in test_join /home/marciorj/Documents/projects/libfunctools/str_join.c:73
    #3 0xaaaac5a822ec in main /home/marciorj/Documents/projects/libfunctools/str_join.c:85
    #4 0xffffa18473f8 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #5 0xffffa18474c8 in __libc_start_main_impl ../csu/libc-start.c:392
    #6 0xaaaac5a8156c in _start (/home/marciorj/Documents/projects/libfunctools/build/str_join+0x156c)

SUMMARY: AddressSanitizer: 9 byte(s) leaked in 1 allocation(s).
make: *** [Makefile:34: test] Error 1
```

Other files have similar errors.

--- a/functools.c
+++ b/functools.c
@@ -49,8 +49,7 @@ ObjList map_data(MapDataFn fn, const void* input, size_t el_len, size_t el_count
     size_t j = 0; // index of the current element
     unsigned char* ix = (unsigned char*)input;
     while (j < el_count) {
-        out[j] = malloc(el_len);
-        memcpy(out[j], fn((void*)&ix[j * el_len], j), el_len);
+        out[j] = fn((void*)&ix[j * el_len], j);
         j++;
     }
     out[j] = NULL;
@@ -67,8 +66,11 @@ void* reduce_data(ReduceDataFn fn, const void* input,
     }
     size_t j = 0; // index of the current element
     unsigned char* ix = (unsigned char*)input;
+    void* tmp;
     while (j < el_count) {
-        init = fn(init, (void*)&ix[j * el_len], j);
+        tmp = fn(init, (void*)&ix[j * el_len], j);
+        free(init);
+        init = tmp;
         j++;
     }
     return init;
@@ -118,8 +120,11 @@ void* reduce(ReduceDataFn fn, ObjList input, void* init) {
         return NULL;
     }
     size_t i = 0;
+    void* tmp;
     for (i = 0; input[i] != NULL; i++) {
-        init = fn(init, input[i], i);
+        tmp = fn(init, input[i], i);
+        free(init);
+        init = tmp;
     }
     return init;
 }
@@ -246,8 +251,8 @@ void test_map() {
     assert(*output[3] == 64);
     assert(*output[4] == 125);
     assert(output[5] == NULL);
-    free(output); // let next free_list() free the rest
-    free_list((ObjList)input, 5);
+    free_list((ObjList)output, 0);
+    free_list((ObjList)input, 0);
 }
 
 
@@ -267,7 +272,7 @@ void test_reduce() {
     int* output = (int*)reduce(&sum, (ObjList)input, NULL);
     assert(*output == 15);
     free(output);
-    free_list((ObjList)input, 5);
+    free_list((ObjList)input, 0);
 }
 
 

--- a/str_join.c
+++ b/str_join.c
@@ -72,6 +72,7 @@ void test_join() {
     const char* list7[] = { "a", "b", "", NULL };
     result = str_join(list7, "123");
     assert(strcmp(result, "a123b123") == 0);
+    free(result);
 
     const char* list8[] = { "", "b", "c", NULL };
     result = str_join(list8, "123");

